File: 2022/app/Main12.hs
module Main where

import Data.Char
import Data.Sort
import Debug.Trace
import qualified Data.Map as M
import qualified Data.Set as S

type Node = (Int, Int)
data Distance a = Dist a | Infinity deriving (Show, Eq)

instance (Ord a) => Ord (Distance a) where
  Infinity <= Infinity = False
  Infinity <= _ = False
  _ <= Infinity = True
  Dist a <= Dist b = a <= b

addDist :: (Num a) => Distance a -> Distance a -> Distance a
addDist (Dist a) (Dist b) = Dist (a + b)
addDist _ _ = Infinity

type Grid = M.Map Node Int
type Distances = M.Map Node (Distance Int)
type Unvisited = S.Set Node

height :: Char -> Int
height c
  | c == 'S' = 0
  | c == 'E' = 25
  | otherwise = ord c - ord 'a'

neighbors :: Grid -> Node -> [Node]
neighbors g n@(x, y) = filter valid [(x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)]
  where
    valid n' = M.member n' g && (g M.! n) - (g M.! n') <= 1

findNext :: Unvisited -> Distances -> Node
findNext u = fst . head . sortOn snd . filter (\(k,_) -> S.member k u) . M.toList

solve :: Unvisited -> Distances -> Grid -> (Unvisited, Distances)
solve u d g
 | u == S.empty = (u, d)
 | otherwise =
     let n = findNext u d
         u' = S.delete n u
         newDist = addDist (d M.! n) (Dist 1)
         ns = neighbors g n
         d' = foldl (\d n' -> M.insert n' (min newDist (d M.! n')) d) d ns
     in solve u' d' g

main :: IO ()
main = do
  content <- readFile "12-input-alex.txt"
--  mapM_ putStrLn $ lines content

  let grid = concatMap (\(y, line) -> map (\(x, c) -> ((x, y), c)) $ zip [0..] line) $ zip [0..] $ lines content
--  print grid

  let start = fst . head . filter (\(_, c) -> c == 'S') $ grid
  let end = fst . head . filter (\(_, c) -> c == 'E') $ grid

  let heights = M.fromList $ fmap (fmap height) grid
--  print heights

  let unvisited = S.fromList $ M.keys heights
--  print unvisited

  let distances = M.insert end (Dist 0) . M.fromList . fmap (flip (,) (Infinity :: Distance Int)) $ M.keys heights
--  print distances

  let (_, distances') = solve unvisited distances heights



  print $ distances' M.! start
  print $ snd . head . sortOn snd . filter (\(k,_) -> heights M.! k == 0) . M.toList $ distances'

File: 2022/src/main12.ts
import fs from "fs";


function height(c: string) {
    switch (c) {
        case "S":
            c = 'a';
            break;
        case "E":
            c = 'z';
            break;
    }

    return c.charCodeAt(0) - 'a'.charCodeAt(0);
}

function neighbors(lines: string[], unvisited: string[], currNode: string) {
    const [x, y] = currNode.split(",").map(x => parseInt(x, 10));

    return [
        [x - 1, y],
        [x + 1, y],
        [x, y - 1],
        [x, y + 1],
    ].filter(([x1, y1]) => x1 >= 0 && y1 >= 0 && x1 < lines[0].length && y1 < lines.length &&
        height(lines[y][x]) - height(lines[y1][x1]) <= 1 &&
        unvisited.indexOf(`${x1},${y1}`) >= 0
    ).map(([x1, y1]) => `${x1},${y1}`)
}

function nextNode(dist: {[k:string]: number}, unvisited: string[]) {
    let min = Infinity;
    let currNode: string | null = null;
    for (let i = 0; i < unvisited.length; i++) {
        const node = unvisited[i];
        // console.log(node, dist[node], min);
        if (dist[node] < min) {
            min = dist[node];
            currNode = node;
        }
    }

    return currNode;
}

function updateDistances(dist: {[k:string]: number}, currNode: string, ns: string[]) {
    for (let i = 0; i < ns.length; i++) {
        const neighbor = ns[i];
        const newDist = dist[currNode] + 1;
        if (newDist < dist[neighbor]) {
            dist[neighbor] = newDist;
        }
    }

    return dist;
}

function visit(lines: string[], dist: {[k:string]: number}, unvisited: string[], currNode: string) {
    unvisited = unvisited.filter(node => node !== currNode);

    const ns = neighbors(lines, unvisited, currNode);

    dist = updateDistances(dist, currNode, ns);

    return {unvisited, dist};
}

function numberWithSpaces(nr: Number) {

    const str = nr.toString();
    if (str.length === 1) {
        return " " + str + " ";
    } else if (str.length === 2) {
        return " " + str;
    } else {
        return str;
    }
}

function main() {
    const file = fs.readFileSync("12-input-alex.txt").toString();

    // console.log(file);
    const lines = file.split("\n");

    let heights: { [k: string]: number } = {};
    let unvisited: string[] = [];
    let dist: { [k: string]: number } = {};

    let start = "";
    let end = "";
    for (let y = 0; y < lines.length; y++) {
        for (let x = 0; x < lines[y].length; x++) {
            unvisited.push(`${x},${y}`);
            dist[`${x},${y}`] = Infinity;
            heights[`${x},${y}`] = height(lines[y][x]);

            if (lines[y][x] === "S") {
                start = `${x},${y}`;
            }

            if (lines[y][x] === "E") {
                end = `${x},${y}`;
            }
        }
    }

    dist[end] = 0;
    while (unvisited.length > 0) {
        const currNode = nextNode(dist, unvisited);

        if (currNode === null) {
            break;
        }

        let {unvisited: u, dist: d} = visit(lines, dist, unvisited, currNode);
        unvisited = u;
        dist = d;
    }

    for (let y = 0; y < lines.length; y++) {
        for (let x = 111; x < lines[y].length; x++) {
            const node = `${x},${y}`;
            if (dist[node] === Infinity) {
                process.stdout.write("    ");
            } else {
                process.stdout.write(numberWithSpaces(dist[node]+1) + " ");
            }
        }
        process.stdout.write("\n");
    }

    for (let y = 0; y < lines.length; y++) {
        for (let x = 111; x < lines[y].length; x++) {
            const node = `${x},${y}`;
            if (dist[node] === Infinity) {
                process.stdout.write("    ");
            } else {
                process.stdout.write("  "+lines[y][x]+" ");
            }
        }
        process.stdout.write("\n");
    }

    console.log(dist[start]);

    let min = Infinity;
    let minNode = "";
    for (let i = 0; i < Object.keys(dist).length; i++) {
        const node = Object.keys(dist)[i];
        if (heights[node] === 0 && dist[node] < min) {
            min = dist[node];
            minNode = node;
        }
    }
    console.log(min);
    // console.log(lines);
    // console.log(unvisited);
    // console.log(dist);
//
}

main();